doc/api/n-api.md

```diff
@@ -4455,6 +4613,57 @@ object `js_object` using `napi_wrap()` and removes the wrapping. If a finalize
 callback was associated with the wrapping, it will no longer be called when the
 JavaScript object becomes garbage-collected.
 
+### napi_type_tag_object
+
+> Stability: 1 - Experimental
+
+<!-- YAML
+added: REPLACEME
+-->
+```c
+napi_status napi_type_tag_object(napi_env env,
+                                 napi_value js_object,
+                                 const void* type_tag);
+```
+
+* `[in] env`: The environment that the API is invoked under.
+* `[in] js_object`: The JavaScript object to be marked.
+* `[in] type_tag`: The tag with which the object is to be marked.
+
+Returns `napi_ok` if the API succeeded.
+
+Associates the value of the `type_tag` pointer with the JavaScript object.
+`napi_check_object_type_tag()` can then be used to compare the tag that was
+attached to the object with one owned by the addon to ensure that the object
+has the right type.
+
+### napi_check_object_type_tag
+
+> Stability: 1 - Experimental
+
+<!-- YAML
+added: REPLACEME
+-->
+```c
+napi_status napi_check_object_type_tag(napi_env env,
+                                       napi_value js_object,
+                                       const void* type_tag,
+                                       bool* result);
+```
```

Comment:
This API could be a bit unhandy if you need to support multiple potential object types, essentially forcing you to check every possible type individually.

Would it make more sense to provide an API to retrieve the type tag?

Reply:
@addaleax I can imagine that such an API might make it too easy to spoof objects. I'm happy to write it, but how about

```c
#define NAPI_TYPE_TAG_NOT_FOUND SIZE_MAX

napi_status napi_object_has_type_tag(napi_env env,
                                     napi_value object,
                                     size_t length,
                                     napi_type_tag* type_tags,
                                     size_t* type_tag_index);
```

I was thinking we could get away with a single `size_t` parameter, à la `argc` from `napi_get_cb_info()`:

https://github.com/nodejs/node/blob/4cee226eb41f2c2e3c71800facc0cbaa2f06ecb5/src/js_native_api.h#L266-L273

but it might be confusing that going in it's a length, and coming out it's an index, so it might be best to stick with two `size_t`s.

Reply:
Yeah, I think that’s okay but also something that could be implemented on top of the current API in a straightforward manner, so I don’t think we need to include it.